File: LP-codes(3rd version)/opr.cpp
#include <bits/stdc++.h>
using namespace std;
bool search(int key, vector<int>& fr)
{
for (int i = 0; i < fr.size(); i++)
if (fr[i] == key)
return true;
return false;
}
int predict(int pg[], vector<int>& fr, int pn, int index)
{
int res = -1, farthest = index;
for (int i = 0; i < fr.size(); i++) {
int j;
for (j = index; j < pn; j++) {
if (fr[i] == pg[j]) {
if (j > farthest) {
farthest = j;
res = i;
}
break;
}
}
if (j == pn)
return i;
}
return (res == -1) ? 0 : res;
}
void optimalPage(int pg[], int pn, int fn)
{
vector<int> fr;
int hit = 0;
for (int i = 0; i < pn; i++) {
if (search(pg[i], fr)) {
hit++;
continue;
}
if (fr.size() < fn)
fr.push_back(pg[i]);
else {
int j = predict(pg, fr, pn, i + 1);
fr[j] = pg[i];
}
}
cout << "No. of hits = " << hit << endl;
int miss=pn-hit;
cout << "No. of misses = " <<miss << endl;
float hit_ratio=(float(hit)/pn)*100;
float miss_ratio=(float(miss)/pn)*100;
cout << "Ratio of misses = " <<miss_ratio << endl;
cout << "Ratio of hits = " <<hit_ratio << endl;
}
int main()
{
int fn;
cout<<"Enter no of farmes : ";
cin>>fn;
int pn;
cout<<"Enter length of string : ";
cin>>pn;
int pg[pn];
cout<<"Enter values : ";
for(int i=0;i<pn;i++)
cin>>pg[i];
optimalPage(pg, pn, fn);
return 0;
}
// Output:
// Enter no of farmes : 3
// Enter length of string : 14
// Enter values : 7 0 1 2 0 3 0 4 2 3 0 3 2 3
// No. of hits = 7
// No. of misses = 7
// Ratio of misses = 50
// Ratio of hits = 50

File: LP-codes (4th version)/PageReplacementStrategies/OPR.java

// Java program to demonstrate optimal page
// replacement algorithm.
import java.io.*;
import java.util.*;

public class OPR {

    // Function to check whether a page exists
    // in a frame or not
    static boolean search(int key, int[] fr) {
        for (int i = 0; i < fr.length; i++)
            if (fr[i] == key)
                return true;
        return false;
    }

    // Function to find the frame that will not be used
    // recently in future after given index in pg[0..pn-1]
    static int predict(int pg[], int[] fr, int pn,
            int index) {
        // Store the index of pages which are going
        // to be used recently in future
        int res = -1, farthest = index;
        for (int i = 0; i < fr.length; i++) {
            int j;
            for (j = index; j < pn; j++) {
                if (fr[i] == pg[j]) {
                    if (j > farthest) {
                        farthest = j;
                        res = i;
                    }
                    break;
                }
            }

            // If a page is never referenced in future,
            // return it.
            if (j == pn)
                return i;
        }

        // If all of the frames were not in future,
        // return any of them, we return 0. Otherwise
        // we return res.
        return (res == -1) ? 0 : res;
    }

    static void optimalPage(int pg[], int pn, int fn) {
        // Create an array for given number of
        // frames and initialize it as empty.
        int[] fr = new int[fn];

        // Traverse through page reference array
        // and check for miss and hit.
        int hit = 0;
        int index = 0;
        for (int i = 0; i < pn; i++) {

            // Page found in a frame : HIT
            if (search(pg[i], fr)) {
                hit++;
                continue;
            }

            // Page not found in a frame : MISS

            // If there is space available in frames.
            if (index < fn)
                fr[index++] = pg[i];

            // Find the page to be replaced.
            else {
                int j = predict(pg, fr, pn, i + 1);
                fr[j] = pg[i];
            }
        }
        System.out.println("No. of hits = " + hit);
        System.out.println("No. of misses = " + (pn - hit));
    }

    // driver function
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter no. of pages: ");
        int n = sc.nextInt();
        int pg[] = new int[n];
        for (int i = 0; i < n; i++) {
            pg[i] = sc.nextInt();
        }
        System.out.println("Enter no. of Frames: ");
        int frames = sc.nextInt();
        optimalPage(pg, n, frames);
    }
}
